### DISCUSSION TITLE
US-specific: End of universal application from 2025?

### DISCUSSION THREAD

**Discussion by eprbell** [eprbell/rp2#134] [General]
Check this article: https://www.jdsupra.com/legalnews/the-action-cryptocurrency-investors-8183288/.

The article is not very clear, but it says that "the concept of averaging your entire digital asset portfolio (sometimes referred to as the “universal wallet approach”) has come to an end." 

I'm not sure what it means with "averaging your entire digital asset portfolio", or with "universal wallet approach". I hope it doesn't mean the concept of universal application, which was discussed in great detail in #79 and that allowed us to reintroduce LIFO and HIFO in RP2.

If anybody has clarity on this, please reply here.

**Comment by eprbell** [comment]
Discussion on Reddit: https://www.reddit.com/r/CryptoTax/comments/1ga0wk5/new_tax_law_usa_forcing_fifo_for_all_transactions/

**Comment by eprbell** [comment]
More detailed information here: https://www.reddit.com/r/CryptoTax/comments/1gbvfic/revenue_procedure_202428_what_you_need_to_know/

It seems that universal application will not be usable starting in 2025...

**Comment by eprbell** [comment]
CC: @macanudo527

**Comment by macanudo527** [comment]
Well, I always thought the universal method was a little forgiving, but something the US did in order to make it easier for people to file their taxes and for them to get paid. Now, they want to make it really hard for us to calculate what we owe them. This will require reworking of the transaction engine I expect. In theory, it shouldn't be that hard, just keeping track of multiple wallets, right?

It'll just require a boatload of testing.

**Reply by eprbell** [reply]
> Well, I always thought the universal method was a little forgiving, but something the US did in order to make it easier for people to file their taxes and for them to get paid. Now, they want to make it really hard for us to calculate what we owe them. This will require reworking of the transaction engine I expect. In theory, it shouldn't be that hard, just keeping track of multiple wallets, right?
> 
> It'll just require a boatload of testing.

It'll definitely require lots of unit tests, but supporting per-wallet accounting won't be trivial. We'll have to hammer in per-wallet queue creation and management before the accounting engine starts and then integrate multiple queues with the accounting engine. I'll have to think a bit more on the design...

**Comment by eprbell** [comment]
I asked some questions on the Reddit thread to clarify the semantics: https://www.reddit.com/r/CryptoTax/comments/1gbvfic/comment/ltvfqy7/?context=3

**Comment by eprbell** [comment]
Closing the discussion: I opened issue #135 to track this.
